### PR TITLE
OpenVX HAL calls disabled for images less than certain threshold

### PR DESCRIPTION
**Pull request by terfendail**
### This pullrequest changes
Disable calls to OpenVX for small image processing in order to avoid slowdown due to call overhead

### CODE REVIEW

File: modules/core/include/opencv2/core/openvx/ovx_defs.hpp [hunk 1]
```diff
@@ -24,6 +24,17 @@ namespace cv{
 namespace ovx{
 // Get common thread local OpenVX context
 CV_EXPORTS_W ivx::Context& getOpenVXContext();
+
+template <int kernel_id> inline bool skipSmallImages(int w, int h)     { return w*h < 3840 * 2160; }
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/openvx/ovx_defs.hpp]: I think, this implementation-dependent checks must not be put to a header file; please, move them to .cpp's.

**Comment by vpisarev** [comment]
@terfendail, what's the status of this PR?

**Comment by terfendail** [comment]
I've tried to move checks to cpp file but found out that it had to be marked with `CV_EXPORTS` in order to be visible for `imgproc` module.
I think it doesn't make sense to add this functions to public interface and also suppose that it could prevent compiler from inlining. Also it will require explicit instantiation of functions for all available kernels.

So at the moment I'v reverted code back to original state and retain tested version on `ovhhal_skipsmall_export` branch

File: modules/core/include/opencv2/core/openvx/ovx_defs.hpp [hunk 2]
```diff
@@ -24,6 +24,17 @@ namespace cv{
...
+template <> inline bool skipSmallImages<VX_KERNEL_MINMAXLOC>(int w, int h) { return w*h < 3840 * 2160; }
+template <> inline bool skipSmallImages<VX_KERNEL_MEDIAN_3x3>(int w, int h) { return w*h < 1280 * 720; }
+template <> inline bool skipSmallImages<VX_KERNEL_GAUSSIAN_3x3>(int w, int h) { return w*h < 320 * 240; }
+template <> inline bool skipSmallImages<VX_KERNEL_BOX_3x3>(int w, int h) { return w*h < 640 * 480; }
+template <> inline bool skipSmallImages<VX_KERNEL_HISTOGRAM>(int w, int h) { return w*h < 2048 * 1536; }
+template <> inline bool skipSmallImages<VX_KERNEL_SOBEL_3x3>(int w, int h) { return w*h < 320 * 240; }
+template <> inline bool skipSmallImages<VX_KERNEL_CANNY_EDGE_DETECTOR>(int w, int h) { return w*h < 640 * 480; }
+template <> inline bool skipSmallImages<VX_KERNEL_ACCUMULATE>(int w, int h) { return w*h < 120 * 60; }
+template <> inline bool skipSmallImages<VX_KERNEL_FAST_CORNERS>(int w, int h) { return w*h < 800 * 600; }
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/openvx/ovx_defs.hpp]: I do not quite understand why the corresponding functions cannot be moved to the respective .cpp files?

Comment (vpisarev) [inline modules/core/include/opencv2/core/openvx/ovx_defs.hpp]: It relates to the particular implementation, right? So, these conditions will likely evolve to something like:

<pre>
#if INTEL_OPENVX
return w*h < constant_A;
#elif AMD_OPENVX
return w*h < constant_B;
#else
...
#endif
</pre>

such code should not pollute header files, it must be put to .cpp

Comment (terfendail) [inline modules/core/include/opencv2/core/openvx/ovx_defs.hpp]: I've moved specializations to respective .cpp files

**Comment by vpisarev** [comment]
thanks! :+1:

**Pull request merged**